Add Contribution guidelines to github

RonEld opened the pull request:
## Description
Add Contribution Guidelines that will be shown in github, when PRs are
made.


## Status
**READY**

## Requires Backporting
NO  


## Additional comments
These are contribution guidelines to the mbed TLS repository, no tests or Changelog needed

Patater commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,40 @@
+
+### Tests
+
+As mentioned, tests that show the correctness of the feature\bug fix should be added to the Pull Request, if not such test exist.  
```
```diff
- the feature\bug fix
+ the feature or bug fix
```

```diff
- the Pull Request, if not such test exist.
+ the pull request, if no such tests already exist.
```

Patater commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,40 @@
+
+### Continuous Integration Tests
+
+Once a PR has been made, the Continuous Integration tests ( CI ) are triggered and run. You should follow the result of the CI tests, and fix failures.  
```
```diff
-  Continuous Integration tests ( CI )
+  continuous integration (CI) tests
```

RonEld commented:
@Patater Thanks for your comments. I modified according to your comments. Most of the fixed sections were copied from the Readme file, so I guess we should modify the grammer errors\ typos in the Readme as well...

@gilles-peskine-arm @andresag01 I fixed according to your comments in the mail, please review

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+ - As with any open source project, contributions will be reviewed by the project team and community and may need some modifications to be accepted.
+ - The contribution should not break API or ABI, unless there is a real justification for that. If there is an API change, the contribution, if accepted, will be merged only when there will be a major release.
+
+### Contributor License Agreement ( CLA )
```
Please remove spaces after `(` and before `)`

andresag01 replied:
Normally, the number of `#` characters at the front means the hierarchy of the document. So for example, this would be element 1.1.1. However, there are no top-level heading, so I would suggest to use only 1 `#` if there is no heading about this. Also, note that the main mbed TLS readme is actually using `-` and `=` characters for the top-level headings:
https://raw.githubusercontent.com/ARMmbed/mbedtls/development/README.md

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+We gratefully accept bug reports and contributions from the community. There are some requirements we need to fulfill in order to be able to integrate contributions:
```
Could you please add a title for this document?

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+- We would ask that contributions conform to [our coding standards](https://tls.mbed.org/kb/development/mbedtls-coding-standards), and that contributions are fully tested before submission.
+- The code should be written in a clean and readable style.
+- The code should be written in a portable generic way, that will benefit the whole community, and not only your own needs.
+- The code should be secure, and will be reviewed in a security point of view as well.
```
Style: `in a security...` -> `from a security...`

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+### Making a Contribution
+
+1. [Check for open issues](https://github.com/ARMmbed/mbedtls/issues) or [start a discussion](https://tls.mbed.org/discussions) around a feature idea or a bug.
+2. Fork the [mbed TLS repository on GitHub](https://github.com/ARMmbed/mbedtls) to start making your changes. As a general rule, you should use the "development" branch as a basis.
```
I would make `development` or `development branch` a link to the development branch in GitHub.

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+1. [Check for open issues](https://github.com/ARMmbed/mbedtls/issues) or [start a discussion](https://tls.mbed.org/discussions) around a feature idea or a bug.
+2. Fork the [mbed TLS repository on GitHub](https://github.com/ARMmbed/mbedtls) to start making your changes. As a general rule, you should use the "development" branch as a basis.
+3. Write a test which shows that the bug was fixed or that the feature works as expected.
+4. Send a pull request (PR) and work with us until it gets merged and published. Contributions may need some modifications, so  a few rounds of review and fixing may be necessary. We will include your name in the ChangeLog :)
```
I would link `pull request (PR)` to the GitHub documentation that explains what a PR is and how to create one.

RonEld replied:
I think that we shouldn't really teach our users github. I would assume that if one want to make a Pull Request, they should know what it is

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+1. All interfaces should be documented through Doxygen. New APIs should introduce Doxygen documentation.
+2. Complex parts in the code should include comments.
+3. If needed, a Readme file is advised
```
Style: missing `.` at end of sentence

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+1. All interfaces should be documented through Doxygen. New APIs should introduce Doxygen documentation.
+2. Complex parts in the code should include comments.
+3. If needed, a Readme file is advised
+4. If a KB article should be added, write this as a comment in the PR description.
```
It is not clear what KB is... Perhaps this should write `knowledgebase (KB)` and be a link to the tls.mbed.org website

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+2. Complex parts in the code should include comments.
+3. If needed, a Readme file is advised
+4. If a KB article should be added, write this as a comment in the PR description.
+5. A Changelog entry should be added for this contribution.
```
Not all fixes require a ChangeLog entry. We should elaborate a little on this.

andresag01 replied:
`Changelog` -> `ChangeLog`

andresag01 replied:
I would also link the ChangeLog to the relevant file in the repository.

andresag01 replied:
I would avoid writing the number sequence on this list. Instead, please use the markdown syntax (https://guides.github.com/features/mastering-markdown/). That is, write `1.` in every list item and this should be displayed correctly in the document viewer.

RonEld replied:
I prefer that all the changes would be added to the ChangeLog, and we decide whether to merge it or not

sbutcher-arm replied:
@RonEld We want to allow contributors to get it right first time as much as possible. I don't think we should ask them to do something that may force a change, and our contributors are pretty smart, so I'd rather leave it to their judgement of how significant a change it is.

RonEld replied:
I agree our contributors are smart, but different people have different views, and even we have our arguments what should be logged in the ChangeLog and what shouldn't

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+5. A Changelog entry should be added for this contribution.
+
+
+ 
```
Please remove trailing spaces at the end of this file

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+3. Write a test which shows that the bug was fixed or that the feature works as expected.
+4. Send a pull request (PR) and work with us until it gets merged and published. Contributions may need some modifications, so  a few rounds of review and fixing may be necessary. We will include your name in the ChangeLog :)
+5. For quick merging, the contribution should be short, and concentrated on a single feature or topic. The larger the contribution is, the longer it would take to review it and merge it.
+6. mbed TLS is release with Apache license, and as such, all the added files should include the Apache license header.
```
I would avoid writing the number sequence on this list. Instead, please use the markdown syntax (https://guides.github.com/features/mastering-markdown/). That is, write `1.` in every list item and this should be displayed correctly in the document viewer.

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+  
+1. [mbedtls-1.3](https://github.com/ARMmbed/mbedtls/tree/mbedtls-1.3)  
+2. [mbedtls-2.1](https://github.com/ARMmbed/mbedtls/tree/mbedtls-2.1)  
+3. [development](https://github.com/ARMmbed/mbedtls/tree/development)  
```
I would avoid writing the number sequence on this list. Instead, please use the markdown syntax (https://guides.github.com/features/mastering-markdown/). That is, write `1.` in every list item and this should be displayed correctly in the document viewer.

andresag01 commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+  
+1. If the contribution is a new feature or enhancement, no backporting is needed.
+2. Bug fixes should be backported to the legacy branches containing these bugs.
+3. Changes in the API do not require backporting. If a bug fix introduced a new API, such as new error codes, the bug fix should be implemented differently in the legacy branch.
```
I would avoid writing the number sequence on this list. Instead, please use the markdown syntax (https://guides.github.com/features/mastering-markdown/). That is, write `1.` in every list item and this should be displayed correctly in the document viewer.

RonEld commented:
Hi @andresag01 I fixed according to most of your comments. There are two comments I don't entirely agree, and answered. Please notice

gertvdijk commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,64 @@
+  
+1. [mbedtls-1.3](https://github.com/ARMmbed/mbedtls/tree/mbedtls-1.3)
+1. [mbedtls-2.1](https://github.com/ARMmbed/mbedtls/tree/mbedtls-2.1)
+1. [development](https://github.com/ARMmbed/mbedtls/tree/development)
```
the development branch is not a legacy branch, I think? Perhaps you mean the latest 2.x branch?

RonEld replied:
Hi @gertvdijk you are correct, the development branch is not a legacy branch. I meant these are the branches that require porting\pushing the change, I'll rephrase

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+Contributor License Agreement (CLA)
+-----------------------------------
+- All contributions, whether large or small require a Contributor's License Agreement (CLA) to be accepted. This is because source code can possibly fall under copyright law and we need your consent to share in the ownership of the copyright.
+- To accept the Contributor’s License Agreement (CLA), individual contributors can do this by creating an mbed account and [accepting the online agreement here with a click through](https://developer.mbed.org/contributor_agreement/). Alternatively, for contributions from corporations, or those that do not wish to create an mbed account, a slightly different agreement can be found [here](https://w ...
```
This needs to be adapted to the new spelling of `mbed` and `ARM`.

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+Coding Standards
+----------------
+- We would ask that contributions conform to [our coding standards](https://tls.mbed.org/kb/development/mbedtls-coding-standards), and that contributions are fully tested before submission.
```
We should detail on what we mean by fully tested here, providing a list of test commands and scripts that should execute successfully, and/or reference the "Continuous Integration" section below (however, the githooks do for example not include `test-ref-configs.pl`).

RonEld replied:
Perhaps we should modify the githooks to include `test-ref-configs.pl`?

hanno-arm replied:
@RonEld `test-ref-configs.pl` runs for a few minutes, and I'd prefer to not have such larger scripts in the githooks. @andresag01 @Patater What do you think? 
I'd prefer to expand the `Continuous Integration` section below to give an explicit list of the things the user should check himself before pushing, like running the tests, adding the githooks, and potentially running other scripts like `test-ref-configs.pl`.

RonEld replied:
@hanno-arm If it's a matter of few minutes, I agree with you and we should add it to the CI. Anyway, it's a different topic unrelated to this PR

hanno-arm replied:
@RonEld I didn't mean to expand the CI, I meant to expand the CI section (or another one that might be more suitable) to include an explicit list of (the things the CI does too and) what _the user_ is supposed to do before opening the PR.

RonEld replied:
@hanno-arm Actually, I think we should expand the CI
I expect the user to run the CI tests, but nonetheless, we should catch this if they don't run the tests

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+Contributor License Agreement (CLA)
+-----------------------------------
+- All contributions, whether large or small require a Contributor's License Agreement (CLA) to be accepted. This is because source code can possibly fall under copyright law and we need your consent to share in the ownership of the copyright.
```
There's a comma missing after "large or small".

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+1. [Check for open issues](https://github.com/ARMmbed/mbedtls/issues) or [start a discussion](https://tls.mbed.org/discussions) around a feature idea or a bug.
+1. Fork the [mbed TLS repository on GitHub](https://github.com/ARMmbed/mbedtls) to start making your changes. As a general rule, you should use the ["development" branch](https://github.com/ARMmbed/mbedtls/tree/development) as a basis.
+1. Write a test which shows that the bug was fixed or that the feature works as expected.
+1. Send a pull request (PR) and work with us until it gets merged and published. Contributions may need some modifications, so  a few rounds of review and fixing may be necessary. We will include your name in the ChangeLog :)
```
There's a double-space at `so  a few rounds`.

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+1. Write a test which shows that the bug was fixed or that the feature works as expected.
+1. Send a pull request (PR) and work with us until it gets merged and published. Contributions may need some modifications, so  a few rounds of review and fixing may be necessary. We will include your name in the ChangeLog :)
+1. For quick merging, the contribution should be short, and concentrated on a single feature or topic. The larger the contribution is, the longer it would take to review it and merge it.
+1. mbed TLS is release with Apache license, and as such, all the added files should include the Apache license header.
```
Typo: `released` instead of `release`, and `under the` instead of `with`?

hanno-arm commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+Tests
+-----
+As mentioned, tests that show the correctness of the feature or bug fix should be added to the Pull Request, if no such tests exist.  
```
`Pull Request` -> `pull request` as above?

RonEld commented:
Hi @hanno-arm I addressed your comments

sbutcher-arm commented:
@hanno-arm / @Patater - Could you please review this and approve if you feel you can? Thanks

Patater commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+It would be highly appreciated if a contribution would be backported to a legacy branch in addition to the [development branch](https://github.com/ARMmbed/mbedtls/tree/development).
+At the moment, the legacy branches are:
+  
+1. [mbedtls-1.3](https://github.com/ARMmbed/mbedtls/tree/mbedtls-1.3)
```
We should probably update this section with the new legacy branches.

gilles-peskine-arm replied:
I'll push a commit for that.

gilles-peskine-arm replied:
done

Patater commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,63 @@
+
+Contributor License Agreement (CLA)
+-----------------------------------
+- All contributions, whether large or small, require a Contributor's License Agreement (CLA) to be accepted. This is because source code can possibly fall under copyright law and we need your consent to share in the ownership of the copyright.
```
Consider removing "possibly".

gilles-peskine-arm replied:
Consider leaving it in, since it's legally correct.

Patater replied:
"Can" already implies "possibly". This change wouldn't change the meaning.

sbutcher-arm replied:
It doesn't change the meaning, only changes the level of doubt someone might have. We don't want doubt here, so better to remove it. I'd also suggest strengthening it as:

> source code falls under copyright law

Patater commented:
Merged via 4b117d9c9
